pilot/pkg/security/authn/v1beta1/policy_applier.go

```diff
@@ -257,24 +266,76 @@ func convertToEnvoyJwtConfig(jwtRules []*v1beta1.JWTRule) *envoy_jwt.JwtAuthenti
 			})
 		}
 		provider.FromParams = jwtRule.FromParams
-
 		jwtPubKey := jwtRule.Jwks
-		if jwtPubKey == "" {
-			var err error
-			jwtPubKey, err = model.GetJwtKeyResolver().GetPublicKey(jwtRule.JwksUri)
-			if err != nil {
-				log.Errorf("Failed to fetch jwt public key from %q: %s", jwtRule.JwksUri, err)
-				// This is a temporary workaround to reject a request with JWT token by using a fake jwks when istiod failed to fetch it.
-				// TODO(xulingqing): Find a better way to reject the request without using the fake jwks.
-				jwtPubKey = createFakeJwks(jwtRule.JwksUri)
+
+		if features.EnableRemoteJwks && jwtRule.JwksUri != "" {
+			// Use remote jwks if jwksUri is non empty. Parse the jwksUri to get the cluster name,
+			// generate the jwt filter config using remoteJwks.
+			// If failed to parse the cluster name, fallback to let istiod to fetch the jwksUri.
+			// TODO: Implement the logic to auto-generate the cluster so that when the flag is enabled,
+			// it will always let envoy to fetch the jwks for consistent behavior.
+			u, _ := url.Parse(jwtRule.JwksUri)
+			hostAndPort := strings.Split(u.Host, ":")
+			host := hostAndPort[0]
+			// TODO: Default port based on scheme ?
+			port := 80
+			if len(hostAndPort) == 2 {
+				var err error
+				if port, err = strconv.Atoi(hostAndPort[1]); err != nil {
+					port = 80 // If port is not specified or there is an error in parsing default to 80.
+				}
 			}
-		}
-		provider.JwksSourceSpecifier = &envoy_jwt.JwtProvider_LocalJwks{
-			LocalJwks: &core.DataSource{
-				Specifier: &core.DataSource_InlineString{
-					InlineString: jwtPubKey,
+			_, cluster, err := security_utils.LookupCluster(push, host, port)
+			if err == nil && len(cluster) > 0 {
+				// This is a case of URI pointing to mesh cluster. Setup Remote Jwks and let Envoy fetch the key.
+				provider.JwksSourceSpecifier = &envoy_jwt.JwtProvider_RemoteJwks{
+					RemoteJwks: &envoy_jwt.RemoteJwks{
+						HttpUri: &core.HttpUri{
+							Uri: jwtRule.JwksUri,
+							HttpUpstreamType: &core.HttpUri_Cluster{
+								Cluster: cluster,
+							},
+						},
+						CacheDuration: &duration.Duration{Seconds: 5 * 60}, // TODO: Make this configurable if needed.
+					},
+				}
+			} else {
+				// Fetch keys from JwksUri and setup as inline key.
```

Comment:
this part is exactly the same as the code below (line 323), can you refactor to share the same code instead of duplicate?

Reply:
Yes. Was in my TODO to do it. Changed now.